diesel/src/migration/mod.rs

```diff
@@ -49,7 +63,27 @@ impl<'a> Migration for &'a Migration {
     fn revert(&self, conn: &SimpleConnection) -> Result<(), RunMigrationsError> {
         (&**self).revert(conn)
     }
+
     fn file_path(&self) -> Option<&Path> {
         (&**self).file_path()
     }
+
+    fn metadata(&self) -> Option<&Metadata> {
+        (&**self).metadata()
+    }
+}
+
+/// Represents metadata associated with a migration.
+///
+/// The format of a migration's metadata is dependent on the migration format
+/// being used.
+///
+/// For Diesel's built in SQL file migrations, metadata is stored in a file
+/// called `metadata.toml`. Diesel looks for a single key, `run_in_transaction`.
+/// By default, all migrations are run in a transaction on SQLite and
+/// PostgreSQL. This behavior can be disabled for a single migration by setting
+/// this to `false`.
+pub trait Metadata {
+    /// Get the metadata at the given key, if present
+    fn get(&self, key: &str) -> Option<Cow<str>>;
```

Comment:
I'm concerned that this may be too limited an interface: I wished to store a list of previous revisions as part of the metadata for example, which means they've have to be double-encoded in the toml file.

Perhaps this "get" method could use [erased_serde](https://github.com/dtolnay/erased-serde) to allow it to support structured data? That way I can specify the type when I access the metadata?

Reply:
Alternatively, to keep things simpler we could just be more opinionated about what format metadata comes in, and have the migration supply a raw TOML/RON/JSON/whatever string.

Reply:
> I'm concerned that this may be too limited an interface: I wished to store a list of previous revisions as part of the metadata for example, which means they've have to be double-encoded in the toml file.

I had the same concern and thought about it for a good while. You absolutely do not need to store it double encoded. You're right that this function will return a TOML string for tables/arrays though, and you'll end up calling `toml::from_str` on the result. There's actually a benefit to doing that, since you can then turn the returned string into a type other than `toml::Value` etc.

> Perhaps this "get" method could use erased_serde to allow it to support structured data? That way I can specify the type when I access the metadata?

I'd prefer not to couple to any single serialization framework here. Especially since that assumes that metadata is serialized in a file, which may not be true for Rust migrations, or migrations which decide to use magic comments, etc.

Reply:
Oh, I see, it's serialising back to TOML before returning if it's not a string 🙁 

What if barrel decides not to use TOML to store its metadata? Then I won't know how to deserialize an array without knowing where a migration came from?

> I'd prefer not to couple to any single serialization framework here.

Serde is the de facto standard for serialization, and diesel already has it as an optional dependency, so this doesn't seem overly risky.

> Especially since that assumes that metadata is serialized in a file, which may not be true for Rust migrations, or migrations which decide to use magic comments, etc.

I don't follow this logic? The only assumption would be that the metadata could be converted to a format which is compatible with serde's object model. For embedding migrations, the internal format would not be exposed, we'd just pick any format which supports the whole of that object model (eg. RON).

Overall I'd rather diesel picked a standardised object model (either serde/RON or JSON) and expected migrations to convert their metadata to that format. Given one of those object models, there's a choice of exposing an API based on serde, or exposing the raw RON or JSON, depending on which you'd rather be coupled to (the latter would allow you to entirely avoid coupling to serde).

Reply:
> What if barrel decides not to use TOML to store its metadata? Then I won't know how to deserialize an array without knowing where a migration came from?

I was sort of assuming that if you're caring about specific metadata keys, you're also assuming a migration format, but you're right that isn't ideal. I'll need to think about this more. I was about to say we could have `get_map` and `get_array` as separate functions, but there's no type we can really return from `get_array` unless we have an enum. Perhaps we should just have a *very* naive enum representation of `Value(String), Array(Vec<ThisEnum>), Map(HashMap<String, ThisEnum>)`. (This now is making some assumptions about how maps are represented but I think that's probably reasonable.) Either way I do need to give this more thought.

> Serde is the de facto standard for serialization, and diesel already has it as an optional dependency, so this doesn't seem overly risky.

I'm not implying that there's another serialization framework to use instead, just that some types are difficult to represent with Serde, and I'd rather not push it here.

> For embedding migrations, the internal format would not be exposed, we'd just pick any format which supports the whole of that object model (eg. RON).

This is less about embedded migrations, and more about cases where the metadata is entirely computed.